File: sdk/typescript/src/dicom.ts
/**
 * DICOM client for the National MCP-PAI Oncology Trials network.
 *
 * Wraps the trialmcp-dicom MCP server tools:
 *   - dicom_query: Query DICOM studies by patient, modality, date, etc.
 *   - dicom_retrieve: Retrieve DICOM study/series pointers (WADO-RS URIs)
 *
 * All patient references use pseudonyms. Actual DICOM pixel data is
 * accessed via WADO-RS URIs returned by retrieve; the MCP server
 * never transmits raw imaging data over the MCP protocol.
 */

import { TrialMcpClient } from './client';
import { InvalidInputError } from './errors';
import type {
  DicomQueryParams,
  DicomQueryResult,
  DicomStudy,
  DicomRetrieveParams,
  DicomRetrieveResult,
} from './models';

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const SERVER: 'trialmcp-dicom' = 'trialmcp-dicom';

/** Supported DICOM modalities for oncology imaging */
const ONCOLOGY_MODALITIES = [
  'CT',    // Computed Tomography
  'MR',    // Magnetic Resonance
  'PT',    // Positron Emission Tomography
  'NM',    // Nuclear Medicine
  'US',    // Ultrasound
  'DX',    // Digital Radiography
  'CR',    // Computed Radiography
  'MG',    // Mammography
  'XA',    // X-Ray Angiography
  'RTDOSE', // Radiation Therapy Dose
  'RTPLAN', // Radiation Therapy Plan
  'RTSTRUCT', // Radiation Therapy Structure Set
  'RTIMAGE', // Radiation Therapy Image
  'SEG',   // Segmentation
  'SR',    // Structured Report
] as const;

type OncologyModality = typeof ONCOLOGY_MODALITIES[number];

/** DICOM UID pattern (dotted numeric) */
const UID_PATTERN = /^[0-9]+(\.[0-9]+)*$/;

/** DICOM date format (YYYYMMDD or range) */
const DATE_PATTERN = /^\d{8}(-\d{8})?$/;

/** Maximum query results */
const MAX_QUERY_RESULTS = 200;

// ---------------------------------------------------------------------------
// DICOM Client
// ---------------------------------------------------------------------------

export class DicomClient {
  private readonly client: TrialMcpClient;

  constructor(client: TrialMcpClient) {
    this.client = client;
  }

  /**
   * Query DICOM studies matching the given criteria.
   *
   * Supports filtering by patient pseudonym, modality, date range,
   * study UID, and accession number. Results are paginated.
   *
   * @param params - Query parameters for filtering studies
   * @returns Matching studies with pagination metadata
   */
  async query(params: DicomQueryParams = {}): Promise<DicomQueryResult> {
    this.validateQueryParams(params);

    const toolParams: Record<string, unknown> = {};

    if (params.study_instance_uid) {
      toolParams.study_instance_uid = params.study_instance_uid;
    }
    if (params.patient_pseudonym) {
      toolParams.patient_pseudonym = params.patient_pseudonym;
    }
    if (params.modality) {
      toolParams.modality = params.modality;
    }
    if (params.study_date) {
      toolParams.study_date = params.study_date;
    }
    if (params.accession_number) {
      toolParams.accession_number = params.accession_number;
    }
    if (params.series_instance_uid) {
      toolParams.series_instance_uid = params.series_instance_uid;
    }
    if (params.limit !== undefined) {
      toolParams.limit = Math.min(params.limit, MAX_QUERY_RESULTS);
    }

    const result = await this.client.callTool<DicomQueryResult>(
      SERVER,
      'dicom_query',
      toolParams,
    );

    return result.data;
  }

  /**
   * Query studies for a specific patient pseudonym.
   * Convenience method for patient-scoped queries.
   */
  async queryByPatient(
    patientPseudonym: string,
    modality?: string,
  ): Promise<DicomStudy[]> {
    const result = await this.query({
      patient_pseudonym: patientPseudonym,
      modality,
    });
    return result.studies;
  }

  /**
   * Query studies by modality (e.g. CT, MR, PT for PET).
   */
  async queryByModality(
    modality: string,
    dateRange?: string,
  ): Promise<DicomStudy[]> {
    const result = await this.query({
      modality,
      study_date: dateRange,
    });
    return result.studies;
  }

  /**
   * Query radiation therapy plans and structures.
   * Common in oncology for treatment planning verification.
   */
  async queryRadiationTherapy(patientPseudonym: string): Promise<DicomStudy[]> {
    const rtModalities = ['RTPLAN', 'RTSTRUCT', 'RTDOSE', 'RTIMAGE'];
    const results: DicomStudy[] = [];

    for (const modality of rtModalities) {
      const queryResult = await this.query({
        patient_pseudonym: patientPseudonym,
        modality,
      });
      results.push(...queryResult.studies);
    }

    return results;
  }

  /**
   * Retrieve a DICOM study/series pointer (WADO-RS URI).
   *
   * Returns a URI that can be used to access the actual DICOM data
   * through a WADO-RS compliant server. The MCP server performs
   * authorization checks before returning the pointer.
   *
   * @param params - Retrieve parameters specifying the target study/series
   * @returns WADO-RS URI and metadata for accessing the DICOM data
   */
  async retrieve(params: DicomRetrieveParams): Promise<DicomRetrieveResult> {
    this.validateRetrieveParams(params);

    const result = await this.client.callTool<DicomRetrieveResult>(
      SERVER,
      'dicom_retrieve_pointer',
      {
        study_instance_uid: params.study_instance_uid,
        series_instance_uid: params.series_instance_uid,
        sop_instance_uid: params.sop_instance_uid,
        transfer_syntax: params.transfer_syntax,
      },
    );

    return result.data;
  }

  /**
   * Retrieve pointers for all series within a study.
   */
  async retrieveStudySeries(studyUid: string): Promise<DicomRetrieveResult> {
    return this.retrieve({ study_instance_uid: studyUid });
  }

  /**
   * Retrieve a specific series within a study.
   */
  async retrieveSeries(
    studyUid: string,
    seriesUid: string,
  ): Promise<DicomRetrieveResult> {
    return this.retrieve({
      study_instance_uid: studyUid,
      series_instance_uid: seriesUid,
    });
  }

  /**
   * Check if a study exists by querying for its UID.
   */
  async studyExists(studyUid: string): Promise<boolean> {
    try {
      const result = await this.query({ study_instance_uid: studyUid, limit: 1 });
      return result.total_count > 0;
    } catch {
      return false;
    }
  }

  // -------------------------------------------------------------------------
  // Validation
  // -------------------------------------------------------------------------

  private validateQueryParams(params: DicomQueryParams): void {
    if (params.study_instance_uid && !UID_PATTERN.test(params.study_instance_uid)) {
      throw new InvalidInputError(
        `Invalid DICOM Study Instance UID format: '${params.study_instance_uid}'`,
        {
          server: SERVER,
          tool: 'dicom_query',
          details: { field: 'study_instance_uid', expectedPattern: 'Dotted numeric UID' },
        },
      );
    }

    if (params.series_instance_uid && !UID_PATTERN.test(params.series_instance_uid)) {
      throw new InvalidInputError(
        `Invalid DICOM Series Instance UID format: '${params.series_instance_uid}'`,
        {
          server: SERVER,
          tool: 'dicom_query',
          details: { field: 'series_instance_uid', expectedPattern: 'Dotted numeric UID' },
        },
      );
    }

    if (params.study_date && !DATE_PATTERN.test(params.study_date)) {
      throw new InvalidInputError(
        `Invalid DICOM date format: '${params.study_date}'. Use YYYYMMDD or YYYYMMDD-YYYYMMDD`,
        {
          server: SERVER,
          tool: 'dicom_query',
          details: { field: 'study_date', expectedPattern: 'YYYYMMDD or YYYYMMDD-YYYYMMDD' },
        },
      );
    }

    if (params.limit !== undefined && (params.limit < 1 || params.limit > MAX_QUERY_RESULTS)) {
      throw new InvalidInputError(
        `Query limit must be between 1 and ${MAX_QUERY_RESULTS}, got ${params.limit}`,
        { server: SERVER, tool: 'dicom_query', details: { field: 'limit' } },
      );
    }
  }

  private validateRetrieveParams(params: DicomRetrieveParams): void {
    if (!params.study_instance_uid) {
      throw new InvalidInputError(
        'study_instance_uid is required for DICOM retrieve',
        { server: SERVER, tool: 'dicom_retrieve_pointer', details: { field: 'study_instance_uid' } },
      );
    }

    if (!UID_PATTERN.test(params.study_instance_uid)) {
      throw new InvalidInputError(
        `Invalid Study Instance UID: '${params.study_instance_uid}'`,
        {
          server: SERVER,
          tool: 'dicom_retrieve_pointer',
          details: { field: 'study_instance_uid', expectedPattern: 'Dotted numeric UID' },
        },
      );
    }

    if (params.series_instance_uid && !UID_PATTERN.test(params.series_instance_uid)) {
      throw new InvalidInputError(
        `Invalid Series Instance UID: '${params.series_instance_uid}'`,
        {
          server: SERVER,
          tool: 'dicom_retrieve_pointer',
          details: { field: 'series_instance_uid', expectedPattern: 'Dotted numeric UID' },
        },
      );
    }
  }
}

File: sdk/python/trialmcp_client/dicom.py
"""DICOM sub-client for imaging operations.

Wraps the two DICOM server tools: ``dicom_query`` and ``dicom_retrieve``.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, Any

from .config import ServerName
from .exceptions import raise_for_error
from .models import DICOMQueryResult, DICOMRetrieveResult, DICOMStudy

if TYPE_CHECKING:
    from .client import TrialMCPClient

logger = logging.getLogger("trialmcp.dicom")

_SERVER = ServerName.DICOM


class DICOMClient:
    """Sub-client for the ``trialmcp-dicom`` MCP server.

    Provides typed, async wrappers around DICOM query and retrieve
    operations.  All patient identifiers in results are pseudonymised
    by the server.

    Parameters
    ----------
    parent:
        The :class:`TrialMCPClient` that owns this sub-client.
    """

    def __init__(self, parent: TrialMCPClient) -> None:
        self._client = parent

    # ------------------------------------------------------------------
    # dicom_query
    # ------------------------------------------------------------------

    async def query(
        self,
        *,
        patient_pseudonym: str | None = None,
        study_instance_uid: str | None = None,
        modality: str | None = None,
        study_date_from: str | None = None,
        study_date_to: str | None = None,
        accession_number: str | None = None,
        limit: int = 50,
    ) -> DICOMQueryResult:
        """Query for DICOM studies matching the given criteria.

        At least one filter parameter must be provided.

        Parameters
        ----------
        patient_pseudonym:
            HMAC-SHA256 pseudonymised patient identifier.
        study_instance_uid:
            Specific DICOM Study Instance UID to look up.
        modality:
            Imaging modality filter (e.g. ``"CT"``, ``"MR"``, ``"PT"``).
        study_date_from:
            Earliest study date (ISO 8601 date string).
        study_date_to:
            Latest study date (ISO 8601 date string).
        accession_number:
            Accession number filter.
        limit:
            Maximum number of studies to return.

        Returns
        -------
        DICOMQueryResult:
            Matching DICOM studies.

        Raises
        ------
        InvalidInputError:
            If no filter parameters are provided.
        """
        params: dict[str, Any] = {"limit": limit}
        if patient_pseudonym:
            params["patient_pseudonym"] = patient_pseudonym
        if study_instance_uid:
            params["study_instance_uid"] = study_instance_uid
        if modality:
            params["modality"] = modality
        if study_date_from:
            params["study_date_from"] = study_date_from
        if study_date_to:
            params["study_date_to"] = study_date_to
        if accession_number:
            params["accession_number"] = accession_number

        result = await self._client.call_tool(_SERVER, "dicom_query", params)
        raise_for_error(result)
        return DICOMQueryResult.from_dict(result)

    # ------------------------------------------------------------------
    # dicom_retrieve
    # ------------------------------------------------------------------

    async def retrieve(
        self,
        *,
        study_instance_uid: str,
        series_instance_uid: str | None = None,
        sop_instance_uid: str | None = None,
        transfer_syntax: str | None = None,
        destination: str | None = None,
    ) -> DICOMRetrieveResult:
        """Retrieve DICOM objects from the imaging archive.

        Parameters
        ----------
        study_instance_uid:
            The Study Instance UID to retrieve.
        series_instance_uid:
            Optional filter to retrieve a specific series.
        sop_instance_uid:
            Optional filter to retrieve a specific SOP instance.
        transfer_syntax:
            Requested DICOM transfer syntax UID.
        destination:
            Optional storage destination path or identifier.

        Returns
        -------
        DICOMRetrieveResult:
            Result including counts and storage path.

        Raises
        ------
        NotFoundError:
            If the study does not exist.
        """
        params: dict[str, Any] = {
            "study_instance_uid": study_instance_uid,
        }
        if series_instance_uid:
            params["series_instance_uid"] = series_instance_uid
        if sop_instance_uid:
            params["sop_instance_uid"] = sop_instance_uid
        if transfer_syntax:
            params["transfer_syntax"] = transfer_syntax
        if destination:
            params["destination"] = destination

        result = await self._client.call_tool(_SERVER, "dicom_retrieve", params)
        raise_for_error(result)
        return DICOMRetrieveResult.from_dict(result)

    # ------------------------------------------------------------------
    # Convenience helpers
    # ------------------------------------------------------------------

    async def query_by_patient(
        self,
        patient_pseudonym: str,
        *,
        modality: str | None = None,
    ) -> list[DICOMStudy]:
        """Query all studies for a pseudonymised patient.

        Parameters
        ----------
        patient_pseudonym:
            The HMAC-SHA256 pseudonymised patient identifier.
        modality:
            Optional modality filter.

        Returns
        -------
        list[DICOMStudy]:
            All matching DICOM studies.
        """
        result = await self.query(
            patient_pseudonym=patient_pseudonym,
            modality=modality,
        )
        return result.studies

    async def retrieve_study(
        self,
        study_instance_uid: str,
    ) -> DICOMRetrieveResult:
        """Retrieve all series and instances for a complete study.

        Parameters
        ----------
        study_instance_uid:
            The Study Instance UID.

        Returns
        -------
        DICOMRetrieveResult:
            Retrieval result with counts and storage location.
        """
        return await self.retrieve(study_instance_uid=study_instance_uid)

    async def query_and_retrieve(
        self,
        *,
        patient_pseudonym: str,
        modality: str | None = None,
    ) -> list[DICOMRetrieveResult]:
        """Query for studies and retrieve all matching ones.

        Combines ``dicom_query`` and ``dicom_retrieve`` into a single
        workflow.

        Parameters
        ----------
        patient_pseudonym:
            The pseudonymised patient identifier.
        modality:
            Optional modality filter.

        Returns
        -------
        list[DICOMRetrieveResult]:
            Retrieval results for each matching study.
        """
        import asyncio

        query_result = await self.query(
            patient_pseudonym=patient_pseudonym,
            modality=modality,
        )

        if not query_result.studies:
            return []

        retrieve_tasks = [
            self.retrieve(study_instance_uid=study.study_instance_uid)
            for study in query_result.studies
        ]
        return list(await asyncio.gather(*retrieve_tasks))

    async def find_study_by_uid(
        self,
        study_instance_uid: str,
    ) -> DICOMStudy | None:
        """Look up a single study by its UID.

        Parameters
        ----------
        study_instance_uid:
            The Study Instance UID.

        Returns
        -------
        DICOMStudy | None:
            The study if found, or ``None``.
        """
        result = await self.query(study_instance_uid=study_instance_uid)
        if result.studies:
            return result.studies[0]
        return None
